Issue#1762: Optional Bouncing Dock Icon

phillipamann opened the pull request:
Here is the PR from the discussion on #1762 .

<!-- You can remove this first section if you have contributed before -->

### First time contributor checklist:
- [x] I have read the [README](https://github.com/WhisperSystems/Signal-Desktop/blob/master/README.md) and [Contributor Guidelines](https://github.com/WhisperSystems/Signal-Desktop/blob/master/CONTRIBUTING.md)
- [x] I have signed the [Contributor Licence Agreement](https://whispersystems.org/cla/)


### Contributor checklist:
- [x] My commits are in nice logical chunks with [good commit messages](http://chris.beams.io/posts/git-commit/)
- [x] My changes are [rebased](https://medium.freecodecamp.org/git-rebase-and-the-golden-rule-explained-70715eccc372) on the latest [`development`](https://github.com/WhisperSystems/Signal-Desktop/tree/development) branch
- [x] My changes pass 100% of [local tests](https://github.com/WhisperSystems/Signal-Desktop/blob/master/CONTRIBUTING.md#tests)
- [x] My changes are ready to be shipped to users


### Description
My PR is addressed in the discussion of #1762. I added a new check box and added new behavior to the check box to make sure that the user can control the dock bouncing on OS X.

I tested this manually. I used two clients and I sent a message from one client to the other on my desktop. Each client had a different profile. I tested the behavior of the application with the checkbox set to true and set to false to verify the dock bouncing behavior worked as intended. When the checkbox was checked, the dock bounced. When the checkbox wasn't checked, the dock did not bounce. The notification icon displayed showing how many messages were received regardless of the dock bouncing. If you hide your dock, you will no longer be distracted.

I didn't write any new tests but all tests were passed.
My OS is OS X 10.13.3.
I did not use any devices.

phillipamann commented:
I see there are some conflicts. I'll fix these in the next day or two.

gasi commented:
Thank you for your contribution, @phillipamann. I am fairly new to the project, so I caught up with #1762 and did some of my own research.

Based on the documentation, it looks like macOS supports dock icon bounce, Windows has flashing the window (`win.flashFrame`), and Linux has neither. I confirmed this with Slack, another popular messaging app that is based on Electron.

### Recommendation

Would you willing to make this feature platform specific and have the appropriate wording for this setting?

- macOS: **Bounce Signal’s icon when receiving a notification**
- Windows: **Flash window when receiving a notification**
- Linux: n/a (hide setting)

To make it more generic, we may rename `dockflashNotification*` to `appIconNotificationAnimation*` or similar.

---

### Reference: Slack
#### macOS

<img width="656" alt="slack_-_fiftythree_family" src="https://user-images.githubusercontent.com/49583/37104696-5b2cc06a-21fb-11e8-8090-e06a716b125b.png">

#### Windows

![pasted_image_3_7_18__11_32_am](https://user-images.githubusercontent.com/49583/37104664-4782e62a-21fb-11e8-84dd-e4125017334a.png)

---

### References
- `app.dock.bounce` macOS only: https://github.com/electron/electron/blob/8c138e74bea6f7be884d750d9c0771a636ba7cf9/docs/api/app.md#appdockbouncetype-macos
- `win.flashFrame` Windows only: https://github.com/electron/electron/blob/a06a1fa26252179209a44378fd550b67d60998a7/docs/tutorial/desktop-environment-integration.md#flash-frame-windows
- Similar discussion for *Mattermost*, an open source Slack alternative: https://github.com/mattermost/desktop/issues/514

phillipamann commented:
Thank you for the very detailed suggestion. I will implement everything you said. I’ll make everything platform specific and I’ll see what I can do to mirror Slack’s behavior. I’ll work on this suggestion this weekend.

phillipamann commented:
Hello, I am very slammed in life but I am going to try and finish all of the suggestions and update the PR now/today.

phillipamann commented:
Hello @gasi, I have almost all of the plan done. The addition of checking the different OS and then offering different texts offers some complexity. I have spent a few hours getting familiar with what I need to do and mapping out where and what I'll add. I'll make the changes tomorrow and do the testing. I have one question though. I still cannot get my local testing environment to work. Can you please reply back so that I can connect to the staging environment to test the dock behavior on my Mac. We'll also need to have someone verify these changes work on Windows and Linux too.

gasi-signal commented:
@phillipamann Thanks for the update. What are the errors you are getting when connecting to the staging environment? Could you share a debug log (`View > Debug Log`)? Alternatively, try connecting to the production environment by creating a new profile configuration in `Signal-Desktop/config`, e.g. `Signal-Desktop/config/local-philipp-development`

```json
{
  "serverUrl": "https://textsecure-service.whispersystems.org",
  "cdnUrl": "https://cdn.signal.org",
  "storageProfile": "philipp-development"
}
```

Hope this helps and I look forward to your PR. If you could set up a VM for testing on Linux and Windows that’d be great but if that’s too hard, I can verify it on those two platforms.

phillipamann commented:
Hello @gasi-signal, I left the cafe yesterday before I could reply to you. I'm back today with the hopeful intention to finish but we'll see. I'll post the debug log for stage now. Secondly, it's a bit too much work right now to set up the VM for Linux and Windows... Let me finish the code first...

phillipamann commented:
Here is the debug log: https://debuglogs.org/5230f87da63fb1f06b31a9e5fe6acaaf7e0e3d32711350df33eebc8807c1e292

gasi-signal commented:
@phillipamann Thanks for the debug logs. I probably won’t get to debug this soon, so I recommend you connect to production as described above to unblock yourself, ok?

phillipamann commented:
Sorry for the snail pace of development. My life is kind of chaotic. I moved to Israel. As I was working and ideally finishing last week, the cafe kicked me out because of a religious holiday I had no idea about and this week I moved because my apartment's water and power failed. I know it sounds like excuses but what can you do. Working now! I think I can finally finish this thing.

phillipamann commented:
I am leaving this comment here if we need to make changes to the Windows flash behavior: https://stackoverflow.com/questions/45504387/electron-win-flashframe-method-when-app-is-minimized

phillipamann commented:
OK, all changes are done for Windows and Mac. I will test Mac now but I need help testing Windows and Linux. I'll let you know if I encounter issues with Mac. I still have to get my environment to work.

phillipamann commented:
I'm going to take a break. I have submitted all of the logic and got my stage testing environment to work but I have hit a snag and maybe you can help me @gasi-signal or @scottnonnenberg-signal? Line 85 and line 92 of settings_view.js are causing me issues when I open preferences. Specifically, the error I see is 

`{"name":"log","hostname":"MacBook-Pro.local","pid":6451,"level":30,"msg":"Model caught error triggering showSettings event: TypeError: OS.isMac is not a function\n    at child.initialize (file://[REDACTED]/Signal-Desktop/js/views/settings_view.js:85:20)\n    at child.Backbone.View (file://[REDACTED]/Signal-Desktop/js/components.js:20767:21)\n    at child.constructor (file://[REDACTED]/Signal-Desktop/js/views/whisper_view.js:28:27)\n    at new child (file://[REDACTED]/Signal-Desktop/js/components.js:21332:41)\n    at child.showSettings (file://[REDACTED]/Signal-Desktop/js/views/inbox_view.js:237:20)\n    at Object.Whisper.events.on (file://[REDACTED]/Signal-Desktop/js/background.js:207:31)\n    at triggerEvents (file://[REDACTED]/Signal-Desktop/js/reliable_trigger.js:60:39)\n    at Object.trigger (file://[REDACTED]/Signal-Desktop/js/reliable_trigger.js:120:17)\n    at EventEmitter.ipc.on ([REDACTED]/Signal-Desktop/preload.js:70:18)\n    at emitOne (events.js:115:13)","time":"2018-04-14T11:28:40.799Z","v":0}`

I simply want to use the isOS() functions to detect what OS the client is running on. After that, I can test the macOS functionality and hopefully, the Windows functionality will work too.

phillipamann commented:
Hello @gasi-signal or @scottnonnenberg-signal? It has been almost 2 weeks... Any help with this simple bug?

scottnonnenberg-signal commented:
Looks to me like you need to do a `yarn grunt` to compile our new typescript parts of the app.

phillipamann commented:
What about importing the isOS() methods? I have been unable to figure that part out. 

Phillip

> On 26 Apr 2018, at 19:01, Scott Nonnenberg <notifications@github.com> wrote:
> 
> Looks to me like you need to do a yarn grunt to compile our new typescript parts of the app.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

phillipamann commented:
This is really old. I am not a front end programmer at all and I have long since turned off desktop notifications (even before this PR). I have since gotten a job and have much less time to work on stuff like this. I am going to close this. Someone else is free to take this work and apply a few small, probably simple touches to make this a reality.
